docs/dev/allocation.qmd

```diff
@@ -0,0 +1,279 @@
+A variable is named `F_basin_in` if the corresponding basin is supplied by a level demand and `F_basin_out` if consumed by a level demand.
+
+```julia
+# Get the node IDs from the subnetwork for basins that have a level demand
+node_ids_basin = [
+    node_id for
+    node_id in graph[].node_ids[subnetwork_id] if graph[node_id].type == :basin &&
+    has_external_demand(graph, node_id, :level_demand)[1]
+]
+problem[:F_basin_in] =
+    JuMP.@variable(problem, F_basin_in[node_id = node_ids_basin,] >= 0.0)
+problem[:F_basin_out] =
+    JuMP.@variable(problem, F_basin_out[node_id = node_ids_basin,] >= 0.0)
+```
+
+The last set of optimization variables is the flow edges in and out of the buffer of nodes with a flow demand. It is defined in a similar way to the second set of variables.
+
+### Setting up initial optimization constraints
+All the variables are greater and equal to 0. This is set when the variables are added to the optimization problem.
+
+Other constraints are `capacity`, `source_user`, `source`, `flow_conservation`, `fractional_flow`, `basin_outflow`, `flow_buffer_outflow` and `flow_demand_outflow`.
+
+For each set of constraints, a function named `add_constrains_[constraints name]` is created.
+
+Take `add_constraints_user_source` as an example, the nodes that are relevant for the constraints are added to the optimization problem by calling JuMP.\@constraint.
+
+```julia
+node_ids_user = [node_id for node_id in node_ids if node_id.type == NodeType.UserDemand]
+
+problem[:source_user] = JuMP.@constraint(
+    problem,
+    [node_id = node_ids_user],
+    F[(node_id, outflow_id(graph, node_id))] <= 0.0,
+    base_name = "source_user"
+)
+```
+
+# Optimization
+Initialization of the data structure is in `allocation_init.jl`, and updating, running and reading the results is in `allocation_optim.jl`.
+
+## Preparing the optimization problem
+### Setting up the objective function
+The optimization objective is the sum of three quadratic error terms.
+The quadratic terms are defined with `add_objective_term` function.
+
+In `set_objective_priority`, it sets the objective function for a given priority with the following steps:
+
+- First, it treats the subnetworks as user demand nodes and adds the quadratic terms of the main network.
+- Then it loops over all the edges in allocation.
+- Based on the type of the node that the edge is pointing to (user demand or flow demand), it adds the corresponding quadratic terms.
+- Finally, it does the same to the edges that start from a level demand node.
+
+### Setting the constraints and capacities
+In the function `set_initial_values`, the following capacities and demands are initialized:
+
+- Source capacities come from the physical layer
+- Edge capacities derived from the maximum capacities between the connected nodes
+- Basin capacities
+- Buffer capacities
+- User demands fractional return flow
+- Demands either come from the Ribasim model or are set via the BMI
```

Comment:
Sources of information:
- Basin capacities come from the disk of water above the max level set by a level demand node
- Buffer capacities and user demand return flow start at 0